### PR TITLE
fix: complete direct editing on focusout

### LINKED ISSUE #2327: Content of the currently edited element is not being updated / stored if a DOM element outside the Canvas is focused
### Describe the Bug

If you are editing an element (Task, Label, does not matter) and you click outside the canvas, the editor is stuck in a state somewhere in-between. The internal Shape data is not being updated as you are still in the state of editing, altough it is not visible.

### Steps to Reproduce

- Go to https://demo.bpmn.io/new
- Create a new diagram
- Insert a task and start editing its title
- Click on the "Download as SVG image" button in the bottom toolbar while the editing is still active

You will see that the generated SVG does not have the title that you just entered:

#### Editing
Image

#### Downloaded SVG

Image

In our use case we are triggering an auto-save if the diagram changes. We also auto-save if our user closes the editor dialog (via a close button in a header toolbar outside the canvas) and there are changes. But there are no actual changes in the BPMN data as the editing is not being recognized as completed.

### Expected Behavior

Update the internal data and close the editing state if the shape / entire canvas loses focus.

### PR DESCRIPTION
**Pull request by JhaSourav07**
### Proposed Changes
This PR implements synchronous `focusout` completion natively on the `TextBox` component. It ensures that direct editing is saved immediately when focus leaves the active editor (e.g., clicking outside the canvas or switching browser windows), preventing the editor from getting stuck in an "open" state. 

By handling this inside `diagram-js-direct-editing`, we eliminate the need for global canvas capture hacks in downstream libraries like `bpmn-js` and `dmn-js`.

### What Changed
* **Native Focus Handling:** Bound a `focusout` listener directly to the `contenteditable` DOM element.
* **Teardown Guards:** Added `isConnected` checks and a `try/finally` re-entrancy guard to prevent `NotFoundError` crashes and double-completions during programmatic teardown.
* **Test Porting:** Ported the edge-case focus loss test suite from `bpmn-js` directly into the core `DirectEditingSpec.js`.

### Checklist
- [x] Brief textual description of the changes present
- [x] The code adheres to the project conventions
- [x] The code is tested (`npm run all` passes)
- [x] The commit messages adhere to the conventional commits guidelines

**Closes bpmn-io/bpmn-js#2327**
**Refs #54** (Resolves the integration issues that stalled the previous approach).

image

### CODE REVIEW

File: lib/TextBox.js [hunk 1]
```diff
@@ -58,6 +58,7 @@ export default function TextBox(options) {
 
   this.keyHandler = options.keyHandler || function() {};
   this.resizeHandler = options.resizeHandler || function() {};
+  this.focusoutHandler = options.focusoutHandler || function() {};
```
[the remark this answers is not in the document]
Comment (JhaSourav07) [reply lib/TextBox.js:63]: Added the focusoutHandler documentation to the TextBox constructor JSDoc.

File: lib/DirectEditing.js [hunk 2]
```diff
@@ -152,6 +153,34 @@ DirectEditing.prototype._handleResize = function(event) {
...
+  // ignore if focus moves to a child element of the textbox
+  if (event.relatedTarget && this._textbox.parent.contains(event.relatedTarget)) {
+    return;
+  }
+
+  // ignore if the textbox content is already detached from the DOM
+  // to prevent NotFoundError during cleanup
+  if (!event.target || !event.target.isConnected) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (JhaSourav07) [reply lib/DirectEditing.js:167]: Switched the teardown guard to use document.documentElement.contains(target) to ensure compatibility with older browsers that lack Node.isConnected.

File: test/DirectEditingSpec.js [hunk 3]
```diff
@@ -709,6 +709,148 @@ describe('diagram-js-direct-editing', function() {
...
+        document.body.appendChild(externalElement);
+
+        var content = directEditing._textbox.content;
+
+        // when - dispatch a focusout with an external relatedTarget
+        var event = new FocusEvent('focusout', {
+          bubbles: true,
+          relatedTarget: externalElement
+        });
+
```
[the remark this answers is not in the document]
Comment (JhaSourav07) [reply test/DirectEditingSpec.js]: Refactored the test suite to use document.createEvent instead of the FocusEvent constructor to prevent crashes in older test environments

File: test/DirectEditingSpec.js [hunk 4]
```diff
@@ -709,6 +709,136 @@ describe('diagram-js-direct-editing', function() {
...
+        // then - complete should not have been called
+        expect(completeSpy).to.not.have.been.called;
+
+        // manually close since we prevented completion
+        directEditing.cancel();
+      }));
+
+
+      it('should NOT call complete() twice when focusout fires re-entrantly', inject(function(directEditing) {
```
Comment (barmac) [inline test/DirectEditingSpec.js:828]: Can you give an example on when this can be necessary?

**Comment by barmac** [comment]
Thank you for your contribution. Please have a look at one question I left: https://github.com/bpmn-io/diagram-js-direct-editing/pull/65#discussion_r2919224298

**Comment by nikku** [comment]
Closed in favor of https://github.com/bpmn-io/diagram-js-direct-editing/pull/74

**Comment by nikku** [comment]
Thanks for your contribution - we'll ensure https://github.com/bpmn-io/diagram-js-direct-editing/pull/74 lands shortly.
